inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp

```diff
@@ -81,9 +81,13 @@ static void insertDiagonalLayerBetween(InferenceEngine::CNNLayerPtr prevLayer,
     auto diagLayer = std::make_shared<ScaleShiftLayer>(LayerParams({diagName, "ScaleShift", Precision::FP32}));
     IE_ASSERT(diagLayer != nullptr);
 
-    // TODO: diagonal size
-    size_t weightsSize = LayerInfo(prevLayer).has32BOutput() ? weightsSize = nextLayer->outData[0]->getDims().back() :
-                                                               Get2DReshapedData(nextLayer->outData[0], 8)->getDims()[1];
+    auto inputLayer = InferenceEngine::CNNNetPrevLayerSkipCertain(nextLayer, 0, [](InferenceEngine::CNNLayerPtr ptr) {
+        return LayerInfo(ptr).isNonFunctional() || LayerInfo(ptr).isSplit() || LayerInfo(ptr).isConcat();
+    });
```

Comment:
consider to share the lambda with the one in `gna_groups.hpp `

Reply:
added isNonValuesChangable() function into LayerInfo which can be re-used